physics/cu_unified_driver.F90

```diff
@@ -0,0 +1,1197 @@
+
+!$acc kernels
+     do i= its,itf
+      forcing(i,:)=0.
+      forcing2(i,:)=0.
+      ccn_gf(i) = 0.
+      ccn_m(i) = 0.
+
+      ! set aod and ccn
+      if (flag_init .and. .not.flag_restart) then
+        aod_gf(i)=aodc0
+      else
+        if((cactiv(i).eq.0) .and. (cactiv_m(i).eq.0))then
+          if(aodc0>aod_gf(i)) aod_gf(i)=aod_gf(i)+((aodc0-aod_gf(i))*(dt/(aodreturn*60)))
+          if(aod_gf(i)>aodc0) aod_gf(i)=aodc0
+        endif
+      endif
+
+      ccn_gf(i)=max(5., (aod_gf(i)/0.0027)**(1/0.640))
+      ccn_m(i)=ccn_gf(i)
+
+      ccnclean=max(5., (aodc0/0.0027)**(1/0.640))
+
+      hbot(i)  =kte
+      htop(i)  =kts
+      raincv(i)=0.
+      xlandi(i)=real(xland(i))
+!     if(abs(xlandi(i)-1.).le.1.e-3) tun_rad_shall(i)=.15
+!     if(abs(xlandi(i)-1.).le.1.e-3) flux_tun(i)=1.5
+     enddo
+     do i= its,itf
+      mconv(i)=0.
+     enddo
+     do k=kts,kte
+      do i= its,itf
+       omeg(i,k)=0.
+       zu(i,k)=0.
+       zum(i,k)=0.
+       zus(i,k)=0.
+       zd(i,k)=0.
+       zdm(i,k)=0.
+      enddo
+     enddo
+
+     psur(:)=0.01*psuri(:)
+     do i=its,itf
+      ter11(i)=max(0.,ht(i))
+     enddo
+     do k=kts,kte
+      do i=its,ite
+       cnvw(i,k)=0.
+       cnvc(i,k)=0.
+       gdc(i,k,1)=0.
+       gdc(i,k,2)=0.
+       gdc(i,k,3)=0.
+       gdc(i,k,4)=0.
+       gdc(i,k,7)=0.
+       gdc(i,k,8)=0.
+       gdc(i,k,9)=0.
+       gdc(i,k,10)=0.
+       gdc2(i,k,1)=0.
+      enddo
+     enddo
+
+      do k=kts,kte
+         do i=its,ite
+            tmfq(i,k)=tmf(i,k,1)
+         enddo
+      enddo
+
+     ierr(:)=0
+     ierrm(:)=0
+     ierrs(:)=0
+     cuten(:)=0.
+     cutenm(:)=0.
+     cutens(:)=0.
+!$acc end kernels
+     ierrc(:)=" "
+!$acc kernels
+     
+
+     kbcon(:)=0
+     kbcons(:)=0
+     kbconm(:)=0
+
+     ktop(:)=0
+     ktops(:)=0
+     ktopm(:)=0
+
+     xmb(:)=0.
+     xmb_dumm(:)=0.
+     xmbm(:)=0.
+     xmbs(:)=0.
+     xmbs2(:)=0.
+
+     k22s(:)=0
+     k22m(:)=0
+     k22(:)=0
+
+     jmin(:)=0
+     jminm(:)=0
+
+     pret(:)=0.
+     prets(:)=0.
+     pretm(:)=0.
+
+     umean(:)=0.
+     vmean(:)=0.
+     pmean(:)=0.
+
+     cupclw(:,:)=0.
+     cupclwm(:,:)=0.
+     cupclws(:,:)=0.
+
+     cnvwt(:,:)=0.
+     cnvwts(:,:)=0.
+     cnvwtm(:,:)=0.
+
+     hco(:,:)=0.
+     hcom(:,:)=0.
+     hcdo(:,:)=0.
+     hcdom(:,:)=0.
+
+     outt(:,:)=0.
+     outts(:,:)=0.
+     outtm(:,:)=0.
+
+     outu(:,:)=0.
+     outus(:,:)=0.
+     outum(:,:)=0.
+
+     outv(:,:)=0.
+     outvs(:,:)=0.
+     outvm(:,:)=0.
+
+     outq(:,:)=0.
+     outqs(:,:)=0.
+     outqm(:,:)=0.
+
+     outqc(:,:)=0.
+     outqcs(:,:)=0.
+     outqcm(:,:)=0.
+
+     subm(:,:)=0.
+     dhdt(:,:)=0.
+
+     do k=kts,ktf
+      do i=its,itf
+        p2d(i,k)=0.01*p2di(i,k)
+        po(i,k)=p2d(i,k) !*.01
+        rhoi(i,k) = 100.*p2d(i,k)/(287.04*(t2di(i,k)*(1.+0.608*qv2di(i,k))))
+        qcheck(i,k)=qv(i,k)
+        tn(i,k)=t(i,k)!+forcet(i,k)*dt
+        qo(i,k)=max(1.e-16,qv(i,k))!+forceqv(i,k)*dt
```

Comment:
Many schemes define a very small value to use as a zero-threshold to avoid instabilities. This could be done here in a future cleanup.